From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Wed, 6 Apr 2016 01:04:23 -0500
Subject: [PATCH] Option to use vanilla per-world scoreboard coloring on names

This change is basically a bandaid to fix CB's complete and utter lack
of support for vanilla scoreboard name modifications.

In the future, finding a way to merge the vanilla expectations in with
bukkit's concept of a display name would be preferable. There was a PR
for this on CB at one point but I can't find it. We may need to do this
ourselves at some point in the future.

diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index fe2c8001897dc6d66ce0d24ddb1a5d3b79810294..538e6751f3bda77ba32256158d2a6a25025b360c 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -247,4 +247,9 @@ public class PaperWorldConfig {
         grassUpdateRate = Math.max(0, getInt("grass-spread-tick-rate", grassUpdateRate));
         log("Grass Spread Tick Rate: " + grassUpdateRate);
     }
+
+    public boolean useVanillaScoreboardColoring;
+    private void useVanillaScoreboardColoring() {
+        useVanillaScoreboardColoring = getBoolean("use-vanilla-world-scoreboard-name-coloring", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java b/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java
index 60b71e62a33e4c1d9665bd9b95a99e7093de2526..baae7968dc05855e4f12367f6c690f8b24acd4c6 100644
--- a/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java
+++ b/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java
@@ -120,6 +120,7 @@ import net.minecraft.network.play.server.SSpawnMobPacket;
 import net.minecraft.network.play.server.STabCompletePacket;
 import net.minecraft.network.play.server.SWorldSpawnChangedPacket;
 import net.minecraft.potion.Effects;
+import net.minecraft.scoreboard.ScorePlayerTeam;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.CommandBlockLogic;
 import net.minecraft.tileentity.CommandBlockTileEntity;
@@ -1834,7 +1835,16 @@ public class ServerPlayNetHandler implements IServerPlayNetHandler {
                     return;
                 }
 
-                s = String.format(event.getFormat(), event.getPlayer().getDisplayName(), event.getMessage());
+                // Paper Start - (Meh) Support for vanilla world scoreboard name coloring
+                String displayName = event.getPlayer().getDisplayName();
+                if (this.field_147369_b.func_130014_f_().paperConfig.useVanillaScoreboardColoring) {
+                    ITextComponent nameFromTeam = ScorePlayerTeam.func_237500_a_(this.field_147369_b.func_96124_cp(), ((CraftPlayer) player).getHandle().func_200200_C_());
+                    // Explicitly add a RESET here, vanilla uses components for this now...
+                    displayName = new net.md_5.bungee.api.chat.TextComponent(net.md_5.bungee.chat.ComponentSerializer.parse(ITextComponent.Serializer.componentToJson(nameFromTeam))).toLegacyText() + org.bukkit.ChatColor.RESET;
+                }
+
+                s = String.format(event.getFormat(), displayName, event.getMessage());
+                // Paper end
                 field_147367_d.console.sendMessage(s);
                 if (((LazyPlayerSet) event.getRecipients()).isLazy()) {
                     for (Object recipient : field_147367_d.func_184103_al().field_72404_b) {
